### PR TITLE
Add support for Claude3, an interactive build mode, and build policy files

### PR DESCRIPTION
**Pull request by ericzbeard**
This PR enhances the `rain build` command. `rain build --prompt` now allows for model selection. Claude 3 Sonnet and Haiku are big improvements over Claude 2. We also added a fully interactive mode for `build`, since the number of options is starting to get complex. And you can now generate OPA Rego and CloudFormation Guard policy files by describing what you want to enforce.

### CODE REVIEW

File: internal/cmd/build/prompt.go [hunk 1]
```diff
@@ -9,18 +9,162 @@ import (
...
 // prompt invokes bedrock to produce a template based on the prompt
-func prompt(p string) {
-	prompt := fmt.Sprintf("Write an AWS CloudFormation YAML template that builds the following:\n\n%s\n\nDo not include any explanation.\n\nWrite only the content of the YAML file.\n\nOutput valid YAML within <yaml></yaml> tags.", p)
-	config.Debugf("About to invoke bedrock claude2 with prompt: %s", prompt)
-	r, err := bedrock.Invoke(prompt)
-	if err != nil {
-		panic(err)
+func runPrompt(p string) {
+	mid := modelId(model)
+	if guard {
```
Comment (hariprakash-j) [inline internal/cmd/build/prompt.go:14]: just a little thing, instead of if a switch case here would keep it clear even if there are more options in the future

File: internal/cmd/build/build.go [hunk 2]
```diff
@@ -400,4 +449,7 @@ func init() {
 	Cmd.Flags().StringVarP(&outFn, "output", "o", "", "Output to a file")
 	Cmd.Flags().BoolVar(&pklClass, "pkl-class", false, "Output a pkl class based on a resource type schema")
 	Cmd.Flags().BoolVar(&noCache, "no-cache", false, "Do not used cached schema files")
+	Cmd.Flags().BoolVar(&guard, "guard", false, "When using --prompt, output a CloudFormation Guard policy file")
+	Cmd.Flags().BoolVar(&rego, "rego", false, "When using --prompt, output an OPA Rego policy file")
+	Cmd.Flags().StringVar(&model, "model", "claude2", "The ID of the Bedrock model to use for --prompt. Shorthand: claude2, claude3haiku, claude3sonnet")
```
Comment (hariprakash-j) [inline internal/cmd/build/build.go:454]: just sharing an idea to make this a bit for flexible and easier to maintain let me know what you think:

maybe it would be cleaner to have two options, `--model` which is just a shorthand with fixed set of values like in the description here and another option `--model-id` where users would specify the specific id if they wish to use any other model other than the default one. These two flags would work together ie. the `--model-id` would be dependent on `--model` this way we can avoid the check functions `isClaude2()` because the user will specify the family if they wish to use a different model. An other benefit i see here is that it will be easier to add other family of models in the future. 

TLDR; we will have a clearer input from the user by using `--model` and `--model-id`.

**Review by hariprakash-j — CHANGES_REQUESTED** [review]

File: internal/cmd/build/build.go (hunk 2 above)
Comment (ericzbeard) [reply internal/cmd/build/build.go:454]: I thought about that, but since each model requires a very different input, it would do no good to allow a user to specify an arbitrary model id. Maybe when we add more models, or if there are different model versions with subtle differences in behavior, but no difference in input.

File: internal/cmd/build/prompt.go (hunk 1 above)
Comment (ericzbeard) [reply internal/cmd/build/prompt.go:14]: Yeah, maybe having guard and rego as bools is not the most scalable.. Maybe something like --filetype "guard|rego|yaml" ?

Comment (hariprakash-j) [reply internal/cmd/build/prompt.go:14]: yea that looks good, and the default value could be cfn :+1:

File: internal/cmd/build/build.go (hunk 2 above)
Comment (hariprakash-j) [reply internal/cmd/build/build.go:454]: understood, I assumed there are variants of claude-2 and claude-3 in bedrock because of the option to enter an id like `anthropic.claude-3-sonnet-20240229-v1:0` instead of the shorthand, this is why i thought it would be nice to let the user specify it like for instance `--model claude3sonnet --model-id anthropic.claude-3-sonnet-20240229-v1:0` or some other claude3sonnet model version. its fine to keep it simple for now :+1:

File: internal/cmd/build/prompt.go (hunk 1 above)
Comment (ericzbeard) [reply internal/cmd/build/prompt.go]: Done

**Review by hariprakash-j — APPROVED** [review]

**Comment by null93** [comment]
I found 2 issues so far. I can take another looks once resolved.

### Terminal Bell During Option Selection

When running the following command and moving selection up/down. I get the annoying terminal bell flash/sound.

Selection is also indented an extra character because of the emoji.

```
% go run cmd/rain/main.go build
✅ template: Create a CloudFormation template
Use the arrow keys to navigate: ↓ ↑ → ←
Select an option to generate the template
    all: Use all schema properties to output a template with placeholders
  ✅ required: Use required schema properties to output a template with placeholders
    prompt: Use Bedrock and Claude to generate a template based on a prompt
    recommend: Output a vetted, recommended template for a use case
```

### No Profile Flag

I do not configure the default profile on my system. Cannot specify profile. 

```
% go run cmd/rain/main.go build
✅ template: Create a CloudFormation template
✅ prompt: Use Bedrock and Claude to generate a template based on a prompt
✅ claude3sonnet: Claude 3 Sonnet
Describe the architecture you want to see in the template: An autoscaling cluster with a web layer, a jump host, a varnish node, a database node, and an opensearch node.
could not establish AWS credentials; please run 'aws configure' or choose a profile
exit status 1
```

```
raffi@mini rain % go run cmd/rain/main.go build --profile some-profile
Error: unknown flag: --profile
Usage:
  rain build [<resource type>] or <prompt>

Flags:
  -b, --bare                 Produce a minimal template, omitting all optional resource properties
      --debug                Output debugging information
  -h, --help                 help for build
  -j, --json                 Output the template as JSON (default format: YAML)
  -l, --list                 List all CloudFormation resource types with an optional name prefix
      --model string         The ID of the Bedrock model to use for --prompt. Shorthand: claude2, claude3haiku, claude3sonnet (default "claude2")
      --no-cache             Do not used cached schema files
      --no-colour            Disable colour output
      --omit-patches         Omit patches and use the raw schema
  -o, --output string        Output to a file
      --pkl-class            Output a pkl class based on a resource type schema
  -p, --prompt               Generate a template using Bedrock and a prompt
      --prompt-lang string   The language to target for --prompt, CloudFormation YAML (cfn), CloudFormation Guard (guard), Open Policy Agent Rego (rego) (default "cfn")
  -r, --recommend            Output a recommended architecture for the chosen use case
  -s, --schema               Output the raw un-patched registry schema for a resource type

exit status 1
```

**Comment by ericzbeard** [comment]
> Terminal Bell During Option Selection

Yes it's super annoying and I can't find how to disable it!

**Comment by null93** [comment]
@ericzbeard Looks like its a known issue with the library. Checkout https://github.com/manifoldco/promptui/issues/49#issuecomment-573814976. The comment talks about how to solve this with "bellSkipper". Its kind of hacky but if it works, it works.

As a side note, if you are looking for a pretty good TUI library consider https://github.com/charmbracelet/bubbletea. They have some pretty good components, checkout https://github.com/charmbracelet/bubbles for some example gifs.

**Comment by ericzbeard** [comment]
> ### No Profile Flag

This is fixed, but `-p` and `-r` are taken by `--profile` and `--region`, so I had to remove those shortcuts from `--prompt` and `--recommend`. I think this is fine, since we have interactive mode now to make things easier and this is not likely a command that is being used in any automation that would break.

**Comment by null93** [comment]
@ericzbeard Ran into another issue. Looks like I do not have access to any of the models:

```
raffi@mini rain % go run cmd/rain/main.go build --profile staging-management-full
✅ template: Create a CloudFormation template
✅ prompt: Use Bedrock and Claude to generate a template based on a prompt
✅ claude3sonnet: Claude 3 Sonnet
Describe the architecture you want to see in the template: Describe the architecture you want to see in the template: An autoscaling cluster with a web layer, a jump host, a varnish node, a database node, and an opensearch node.
error invoking model: operation error Bedrock Runtime: InvokeModel, https response error StatusCode: 403, RequestID: b8ef1216-bef4-43fd-a689-822832eb904c, AccessDeniedException: You don't have access to the model with the specified model ID.
exit status 1
```
```
raffi@mini rain % go run cmd/rain/main.go build --profile staging-management-full
✅ template: Create a CloudFormation template
✅ prompt: Use Bedrock and Claude to generate a template based on a prompt
✅ claude3haiku: Claude 3 Haiku
Describe the architecture you want to see in the template: Describe the architecture you want to see in the template: An autoscaling cluster with a web layer, a jump host, a varnish node, a database node, and an opensearch node.
error invoking model: operation error Bedrock Runtime: InvokeModel, https response error StatusCode: 403, RequestID: 36f4ead6-9843-484b-ac52-6e25b42ef27f, AccessDeniedException: You don't have access to the model with the specified model ID.
exit status 1
```
```
raffi@mini rain % go run cmd/rain/main.go build --profile staging-management-full
✅ template: Create a CloudFormation template
✅ prompt: Use Bedrock and Claude to generate a template based on a prompt
✅ claude2: Claude 2
Describe the architecture you want to see in the template: Describe the architecture you want to see in the template: An autoscaling cluster with a web layer, a jump host, a varnish node, a database node, and an opensearch node.
error invoking model: operation error Bedrock Runtime: InvokeModel, https response error StatusCode: 403, RequestID: d3694e75-aac2-4a0a-867b-17dc694285b4, AccessDeniedException: You don't have access to the model with the specified model ID.
exit status 1
```

Logged in manually and saw that I need to submit a request with a use case.

image

Is this expected? Do you want me to test this functionality out or should I just skip the prompt feature?

**Comment by ericzbeard** [comment]
> @ericzbeard Looks like its a known issue with the library. Checkout [manifoldco/promptui#49 (comment)](https://github.com/manifoldco/promptui/issues/49#issuecomment-573814976). The comment talks about how to solve this with "bellSkipper". Its kind of hacky but if it works, it works.
> 
> As a side note, if you are looking for a pretty good TUI library consider https://github.com/charmbracelet/bubbletea. They have some pretty good components, checkout https://github.com/charmbracelet/bubbles for some example gifs.

Good find! I used the solution farther down that apparently works better with Windows.

**Comment by ericzbeard** [comment]
> @ericzbeard Ran into another issue. Looks like I do not have access to any of the models:

> Is this expected? Do you want me to test this functionality out or should I just skip the prompt feature?

This is expected. You have to activate those models in your account. They are not free.

**Comment by ericzbeard** [comment]
Thanks @hariprakash-j and @null93 for the detailed feedback and testing! I think this is ready to merge. I will get a release out later today.

**Review by null93 — APPROVED** [review]
I tested the rest of the functionality (other than the prompt feature) and it looks great! Good job!

The only things I would suggest changing is breaking up the `build` command into subcommands. It does not seem consistent with the rest of the cli tool. It seems like the build command does a lot of different things and it switches based on the flag you pass.

The other thing is that terminal bell. For the short time I was testing this it got on my nerves a lot lol.

**Comment by ericzbeard** [comment]
> The only things I would suggest changing is breaking up the `build` command into subcommands. 

Yeah, it's getting chunky. That's why I made the interactive mode, so users won't have to decipher all the params. Next time I add anything here, I will look at a redesign.

**Comment by hariprakash-j** [comment]
> I tested the rest of the functionality (other than the prompt feature) and it looks great! Good job!
> 
> The only things I would suggest changing is breaking up the `build` command into subcommands. It does not seem consistent with the rest of the cli tool. It seems like the build command does a lot of different things and it switches based on the flag you pass.
> 
> The other thing is that terminal bell. For the short time I was testing this it got on my nerves a lot lol.

+1 on making the generative features its own command, something like `rain generate` ?
